cpp/src/arrow/extension_type.h

```diff
@@ -45,6 +45,9 @@ class ARROW_EXPORT ExtensionType : public DataType {
   /// \brief The type of array used to represent this extension type's data
   const std::shared_ptr<DataType>& storage_type() const { return storage_type_; }
 
+  /// \brief Return a reference to the storage type
+  const DataType& storage_type_ref() const override { return *storage_type_; }
```

Comment:
Alternatively, I could change the return type and make `storage_type()` virtual:

```suggestion
  std::shared_ptr<DataType> storage_type() const override { return storage_type_; }
```

Then `DataType::storage_type()` can just call `shared_from_this()`

Reply:
I'm in favor of making `storage_type()` virtual.